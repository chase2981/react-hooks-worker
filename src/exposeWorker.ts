/* eslint-disable @typescript-eslint/no-explicit-any */

export function exposeWorker(func: (data: any) => any) {
  self.onmessage = async (e: MessageEvent) => {
    const r = func(e.data);
    if (r[Symbol.asyncIterator]) {
      for await (const i of r) (self.postMessage as any /* FIXME */)(i);
    } else if (r[Symbol.iterator]) {
      for (const i of r) (self.postMessage as any /* FIXME */)(i);
    } else {
      (self.postMessage as any /* FIXME */)(await r);
    }
  };
}
