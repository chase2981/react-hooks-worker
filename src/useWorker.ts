import {
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';

type State = {
  result?: unknown;
  error?: 'error' | 'messageerror';
};

const initialState: State = {};

export function useWorker(createWorker: () => Worker, input: unknown) {
  const [state, setState] = useState<State>(initialState);
  const worker = useMemo(createWorker, [createWorker]);
  const lastWorker = useRef<Worker>();
  useEffect(() => {
    lastWorker.current = worker;
    let setStateSafe = (nextState: State) => setState(nextState);
    worker.onmessage = (e) => setStateSafe({ result: e.data });
    worker.onerror = () => setStateSafe({ error: 'error' });
    worker.onmessageerror = () => setStateSafe({ error: 'messageerror' });
    const cleanup = () => {
      setStateSafe = () => null; // we should not setState after cleanup.
      worker.terminate();
      setState(initialState);
    };
    return cleanup;
  }, [worker]);
  useEffect(() => {
    if (lastWorker.current) {
      lastWorker.current.postMessage(input);
    }
  }, [input]);
  return state;
}
