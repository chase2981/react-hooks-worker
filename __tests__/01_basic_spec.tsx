import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';

import { useWorker } from '../src/index';

global.Worker = jest.fn();

describe('basic spec', () => {
  it('should have a function', () => {
    expect(useWorker).toBeDefined();
  });

  it('should create a component', () => {
    const createWorker = () => new Worker('./slow_fib.worker');
    const CalcFib: React.FC<{ count: number }> = ({ count }) => {
      const { result, error } = useWorker(createWorker, count);
      if (error) return <div>Error: {error}</div>;
      return <div>Result: {result}</div>;
    };
    const renderer = createRenderer();
    renderer.render(<CalcFib count={5} />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
