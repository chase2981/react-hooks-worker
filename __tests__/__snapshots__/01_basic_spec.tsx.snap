// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic spec should create a component 1`] = `
<div>
  Result: 
</div>
`;
